### PR TITLE
Fix the Leaky Roof: Fail Fast on Post-Transaction Iterator Usage

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Prevent users from using Iterators, Streams, Visitables and Async operations after a transaction has ended.
- Ensure that AtlasDB devs do not accidentally allow users to do this in the future

**Implementation Description (bullets)**:
- Add tests.
- In the cases we have not handled correctly, fix the transaction code.
- Add information to the javadoc indicating our decision.

**Testing (What was existing testing like?  What have you done to improve it?)**:
- Tests added that exceptions are thrown if one tries to use the forbidden incantations.

**Concerns (what feedback would you like?)**:
- Did I miss any key tests?

**Where should we start reviewing?**: SnapshotTransactionTest

**Priority (whenever / two weeks / yesterday)**: 1 week

### CODE REVIEW

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java [hunk 1]
```diff
@@ -78,6 +78,8 @@
      * It is guaranteed that the {@link Map#keySet()} of the returned map has a corresponding element for each of the
      * input {@code rows}, even if there are rows where no columns match the predicate.
      *
+     * Batching visitables must be used strictly within the scope of the transaction.
```
Comment (Jolyon-S) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java:81]: This is copy-and-pasted throughout, but I wonder if you maybe need to be a bit more explicit with what you mean at least once (maybe at the top):
```
Batching visitables, iterables, streams, and futures _must_ be used strictly within the scope of the transaction. Concretely, this means that results must be retrieved before exiting the transaction - in the case of streams and iterators, this means that you must collect the results before exiting; in the case of futures, you must await all asynchronous calls before returning from the transaction
```
Or something along those lines. I know what you mean here, but I have a feeling that users may not.
```

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java [hunk 2]
```diff
@@ -143,6 +149,8 @@
      * ordering of rows is undefined. In case of duplicate rows, the ordering is based on the first occurrence of
      * the row.
      *
+     * Iterators must be used strictly within the scope of the transaction.
```
Comment (Jolyon-S) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java:152]: Also, you haven't added a similar comment on `getAsync`!

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 3]
```diff
@@ -1891,6 +1898,108 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
...
     }
 
+    @Test
+    public void cannotCompleteFutureAfterTransactionCommit() {
+        SettableFuture<Object> blocker = SettableFuture.create();
+        ListenableFuture<Map<Cell, byte[]>> getFuture = txManager.runTaskThrowOnConflict(txn -> {
+            txn.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("jeremy")));
+            return Futures.transformAsync(
+                    blocker,
+                    _unused -> txn.getAsync(TABLE, ImmutableSet.of(TEST_CELL)),
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1908]: This is testing something subtly different if I am not mistaken: this is testing that you cannot complete a transaction, then call a method on that transaction. I think you ideally want a test where you start `getAsync` well within the bounds of the transaction, but that you return that (unfinished) future and use that outside. I don't think this matters except in SerializableTransactions, and I appreciate that this is probably hard to test, but that's the behaviour here that was at fault before, I believe.

**Review by Jolyon-S — COMMENTED** [review]
I like the solution, but I think you're missing some details in the javadoc, and there's a test that I'd like to see (in an ideal world, although it may be hard to write).

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java:81]: Makes sense - yeah I've added this at the top. What you've written is very clear, so I've used that :)

File: atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/transaction/api/Transaction.java:152]: good spot - added!

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1908]: You're right: yeah the getAsync only happens after the transaction has completed here. I added one that handles usage of a future that began within (and completed within) but is used outside.

There isn't really a great way of injecting this, though, without considerably rewiring the SnapshotTransaction :(

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 4]
```diff
@@ -788,12 +788,18 @@ private void extractLocalWritesForRow(
...
+                cache.get()
+                        .getAsync(
+                                tableRef,
+                                cells,
+                                (table, uncached) -> getInternal(
+                                        "getAsync", tableRef, uncached, keyValueService, defaultTransactionService)),
+                input -> {
+                    ensureStillRunning();
+                    return input;
+                },
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:801]: technically, this is the output / result, not the input

Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:801]: Also, let's extract a `scopeToTransaction` method for futures too, then it is consistent and a bit cleaner.

File: atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java [hunk 5]
```diff
@@ -1147,9 +1147,6 @@ public void getSortedColumnsOnlyReadCellsAreCheckedForConflicts() {
         t2.commit();
 
         assertThatCode(t1::commit).doesNotThrowAnyException();
-
-        List<Cell> cells = Streams.stream(sortedColumns).map(Map.Entry::getKey).collect(Collectors.toList());
-        sanityCheckOnSortedCells(rows, cells, cellsWrittenOriginally);
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java:1152]: Can we not just move this before the commit? But still, nice catch!

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 6]
```diff
@@ -1891,6 +1897,121 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
         });
     }
 
+    @Test
+    public void cannotUseCompletedFutureAfterTransactionCommit() throws Exception {
+        ListenableFuture<Map<Cell, byte[]>> getFuture = txManager.runTaskThrowOnConflict(txn -> {
+            txn.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("jolyon")));
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1903]: this is the second time my name features in a test!

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 7]
```diff
@@ -1891,6 +1897,121 @@ public void transactionStillCommittedEvenIfCallbackThrows() {
         });
     }
 
+    @Test
+    public void cannotUseCompletedFutureAfterTransactionCommit() throws Exception {
```
Comment (Jolyon-S) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1901]: [Sanity check]: Do we have tests for normal getAsync usage?

**Review by Jolyon-S — APPROVED** [review]
Looks good - glad we spotted this weird edge case. A couple of nits, then good to go.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 6 above)
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java]: ... I found the first one!

```
        assertThatThrownBy(() -> schema.addTableDefinition(TEST_TABLE_NAME, tableDef))
                .as("Jolyon loves this stuff")
                .isInstanceOf(SafeIllegalArgumentException.class)
                .hasMessage("Caching can only be enabled with the SERIALIZABLE_CELL conflict handler.");
```
